modules/hcp-hc-objects.adoc

```diff
@@ -0,0 +1,295 @@
++
+[source,yaml]
+----
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  creationTimestamp: null
+  name: capi-provider-role
+  namespace: <hosted_cluster_namespace>-<hosted_cluster_name> <1> <2>
+rules:
+- apiGroups:
+  - agent-install.openshift.io
+  resources:
+  - agents
+  verbs:
+  - '*'
+----
++
+<1> Replace `<hosted_cluster_namespace>` with the name of your hosted cluster namespace.
+<2> Replace `<hosted_cluster_name>` with your hosted cluster.
+
+. Create a YAML file with information about the `HostedCluster` object, replacing values as necessary:
++
+[source,yaml]
+----
+apiVersion: hypershift.openshift.io/v1beta1
+kind: HostedCluster
+metadata:
+  name: <hosted_cluster_name> <1>
+  namespace: <hosted_cluster_namespace> <2>
+spec:
+  additionalTrustBundle:
+    name: "user-ca-bundle"
+  olmCatalogPlacement: guest
+  imageContentSources: <3>
+  - source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
+    mirrors:
+    - registry.<dns.base.domain.name>:5000/openshift/release <4>
+  - source: quay.io/openshift-release-dev/ocp-release
+    mirrors:
+    - registry.<dns.base.domain.name>:5000/openshift/release-images <4>
+  - mirrors:
+  ...
+  ...
+  autoscaling: {}
+  controllerAvailabilityPolicy: SingleReplica
+  dns:
+    baseDomain: <dns.base.domain.name> <4>
+  etcd:
+    managed:
+      storage:
+        persistentVolume:
+          size: 8Gi
+        restoreSnapshotURL: null
+        type: PersistentVolume
+    managementType: Managed
+  fips: false
+  networking:
+    clusterNetwork:
+    - cidr: 10.132.0.0/14
+    - cidr: fd01::/48
+    networkType: OVNKubernetes
+    serviceNetwork:
+    - cidr: 172.31.0.0/16
+    - cidr: fd02::/112
+  platform:
+    agent:
+      agentNamespace: <hosted_cluster_namespace>-<hosted_cluster_name> <1> <2>
+    type: Agent
+  pullSecret:
+    name: <hosted_cluster_name>-pull-secret <1>
+  release:
+    image: registry.<dns.base.domain.name>:5000/openshift/release-images:4.x.y-x86_64 <4> <5>
+  secretEncryption:
+    aescbc:
+      activeKey:
+        name: <hosted_cluster_name>-etcd-encryption-key <1>
+    type: aescbc
+  services:
+  - service: APIServer
+    servicePublishingStrategy:
+      nodePort:
+        address: api.<hosted_cluster_name>.<dns.base.domain.name> <1> <4>
+      type: NodePort
+  - service: OAuthServer
+    servicePublishingStrategy:
+      nodePort:
+        address: api.<hosted_cluster_name>.<dns.base.domain.name> <1> <4>
+      type: NodePort
+  - service: OIDC
+    servicePublishingStrategy:
+      nodePort:
+        address: api.<hosted_cluster_name>.<dns.base.domain.name> <1> <4>
+      type: NodePort
+  - service: Konnectivity
+    servicePublishingStrategy:
+      nodePort:
+        address: api.<hosted_cluster_name>.<dns.base.domain.name> <1> <4>
+      type: NodePort
+  - service: Ignition
+    servicePublishingStrategy:
+      nodePort:
+        address: api.<hosted_cluster_name>.<dns.base.domain.name> <1> <4>
+      type: NodePort
```

Comment:
same

Reply:
Right, KubeVirt isn't tested with NodePort, instead we use Route for everything except the APIServier, which we use LoadBalancer for. We need to be compatible with NodePort in the future, there's currently an open bug for this here, https://issues.redhat.com/browse/OCPBUGS-39158

This example wouldn't work for KubeVirt for multiple reasons. NodePort being one of the issues. The platform being set to `Agent` is another issue.

Reply:
Hey @lahinson I agree with David that NodePort is one of the configurations that maybe we don't want to promote as the best or main one. So maybe we can reconfigure the services to be based on LoadBalancer and Routes.

This is how should looks like in the above case:

```
  services:
  - service: APIServer
    servicePublishingStrategy:
      type: LoadBalancer
  - service: OAuthServer
    servicePublishingStrategy:
      type: Route
  - service: OIDC
    servicePublishingStrategy:
      type: Route
  - service: Konnectivity
    servicePublishingStrategy:
      type: Route
  - service: Ignition
    servicePublishingStrategy:
      type: Route
```

We will need to have configure MetalLB in the Management Cluster and also we will configure it that too in the HostedCluster. What do you think? 

Maybe we can remark that NodePort is not the desired standard but you can configure it using MetalLB using this documentation: https://hypershift-docs.netlify.app/recipes/agent/exposing-dataplane-with-metallb/

I can provide the configuration needed for the Management Cluster, the HostedCluster one is in the link mentioned above.